flatfs.go

```diff
@@ -364,13 +368,21 @@ func (fs *Datastore) renameAndUpdateDiskUsage(tmpPath, path string) error {
 	// Rename and add new file's diskUsage. If the rename fails,
 	// it will either a) Re-add the size of an existing file, which
 	// was sustracted before b) Add 0 if there is no existing file.
-	err = os.Rename(tmpPath, path)
+	for i := 0; i < RetryAttempts; i++ {
+		err = os.Rename(tmpPath, path)
+		// if there's no error, or the source file doesn't exist, abort.
+		if err == nil || os.IsNotExist(err) {
+			break
+		}
+		// Otherwise, this could be a transient error due to some other
+		// process holding open one of the files. Wait a bit and then
+		// retry.
+		time.Sleep(time.Duration(i+1) * RetryDelay)
+	}
 	fs.updateDiskUsage(path, true)
```

Comment:
I don't know exactly why we do this, but I believe it's because we're concerned that the rename may not be atomic. That is, we:

1. Decrement the disk usage.
2. Rename.
3. Fix the disk usage _always_.

This will work even if the rename returns an error, but actually updates the target file under the covers.